### ISSUE TITLE
Generated tests are not compatible across architectures

### ISSUE THREAD

**Issue by kornelski** [rust-lang-nursery/rust-bindgen#1213]
### Input C/C++ Header

```C++
struct Foo {
   int bar;
};
```

### Bindgen Invocation

Run on `x86_64-apple-darwin`:

```
$ bindgen input.h
```

Test with: `cargo test --target=i686-apple-darwin`

### Actual Results

```
fn bindgen_test_layout_Foo ( ) …
```

### Expected Results

The tests are valid only on the architecture bindgen was run on. It would be great if bindgen was able to make tests portable, e.g.:

```rust
#[cfg(target_pointer_width = "64")]
fn bindgen_test_layout_Foo ( ) …

#[cfg(target_pointer_width = "32")]
fn bindgen_test_layout_Foo ( ) …
```

**Comment by emilio** [comment]
It's pretty hard to do that, it not only depends on architecture but also on compiler flags and what not...

**Comment by emilio** [comment]
We support not generating tests of course, but that's not quite a solution...

**Comment by fitzgen** [comment]
Best thing to do is either generate bindings for the current target in `build.rs` or generate bindings for each target you plan to support and check them into your version control.

There is a lot more gotchas than just pointer width, and the biggest is `#ifdef`s that change definitions based on target arch or os etc.

**Comment by kornelski** [comment]
IIRC all of the platform-specific `ifdefs` that I had to deal with were only used to get reliable/cross-platform fixed-size types (in codebases that think depending on `stdint.h` is too radical).

**Comment by kornelski** [comment]
I think the minimal fix here would be to add `#[cfg(target_pointer_width = "…")]` to the tests, so that at least they don't give false negatives on other platforms.

**Comment by Dushistov** [comment]
@kornelski 

> For libraries with stable ABI I prefer to generate bindings once
> I think the minimal fix here would be to add `target_arch`
> to the tests

But what about generated bindings by itself?
For example for:

```C
#include <stdint.h>

union Foo {
        void *p;
        uint32_t a;
};
```
`bindgen` generate different code for different platforms:

```
#[repr(C)]
#[derive(Copy, Clone)]
pub union Foo {
    pub p: *mut ::std::os::raw::c_void,
    pub a: u32,
    _bindgen_union_align: u32,
}

#[repr(C)]
#[derive(Copy, Clone)]
pub union Foo {
    pub p: *mut ::std::os::raw::c_void,
    pub a: u32,
    _bindgen_union_align: u64,
}
```

and failed tests indicates that something wrong, and you should generate bindings in `target` directory, if tests don't fail with different `--target=`, then you may assume that all good, while really all bad.

May be leave current behaviour as is, because of it prevents errors and bugs?

**Comment by kornelski** [comment]
In your example I actually don't know why bindgen does it this way, because `::std::mem::align_of::<::Foo>()` is also correct without bindgen's extra field (4 on i686, 8 on x86_64).

I'm working under assumption that `#[repr(C)]` struct used with `::std::os::raw::c_*` types (and raw pointers) is correct for C without bindgen's hacks if C doesn't use extra pragmas for packing structs, custom align or C++ features. Fortunately these assumptions happen to be true very often for small libraries with C ABI.

**Comment by emilio** [comment]
Yeah we could avoid generating the alignment for the union in some cases, I'm happy to accept patches for that as long as there are proper tests.
